bchehraz/react-challenge-project-jan-2022
# Challenge 9 (Bonus): API call refactor

## Changes
_List Changes Introduced by this PR_
1. (from challenge 1) Fixes order form input by changing `event.value` to `event.target.value`
2. (from challenge 4) Fixes authReducer action payload to properly login
3. (from challenge 5) Wires in the logout button in the `nav` component.
4. (from challenge 6) Wires in `PrivateRoute` route guard component
5. (from challenge 8) Wires in Add/Delete functionality
6. Creates new `orderReducer` and `orderActions`, accompanied by a new `get-current-orders` action type, to house all current API actions related to orders (add, edit, delete, and fetching all current orders).
7. Refactors the affected components, in this case: `ordersList`, `viewOrders`, and `order-form`.

## Purpose
_Describe the problem or feature in addition to a link to the issues._

- Link to issue: https://github.com/Shift3/react-challenge-project-jan-2022/issues/9
- Over time, the components will become diluted with various API calls and functionality that make the codebase harder to maintain and more likely for code to be repeated. Since the project is already using Redux, we may as well take advantage to simplify the code, thus making the components more concise, easier to understand, and more maintainable/scalable.

## Approach
_How does this change address the problem?_

- By moving all API calls related to orders into a new actions file `orderActions.js`, we can have a central location where we implement new server calls related to orders, or really any logic that may affect app state related to orders. In addition, `orderReducer` will ensure every time an action related to orders takes place, that the app state reflects this. This gives every component the ability to use any order actions as necessary and the ability to read app state from a centralized location.

## Testing Steps
_How do the users test this change?_

Testing will run through the various order actions: addOrder, editOrder, deleteOrder, and getCurrentOrders.

1. First add several orders via the `/order` page.
2. Then go to the `/view-orders` page to see if the newly added orders are on the screen. This ensures `getCurrentOrders` is functioning and verifies `addOrder` action from the previous step is also working.
3. Now test `deleteOrder` action by clicking the delete button on any of the newly added orders.
4. Now test the `editOrder` action by clicking edit and making a change to an order. Upon update, the list should reflect the newly changed order.

## Learning
_Describe the research stage_

- I would frequently Google any issues I may have ran into while implementing this feature, and at one point referred to this article as it has a useful example using redux (https://nikgrozev.com/2017/09/16/redux-common-types/)

Closes https://github.com/Shift3/react-challenge-project-jan-2022/issues/9
